### ISSUE TITLE
[BUG] [3006.x] saltutil.cmd: "Authentication error occurred." when master is not running as root

### ISSUE THREAD

**Issue by hurzhurz** [saltstack/salt#64398]
**Description**
When the master is running as user 'salt' (not 'root'), saltutil.cmd fails with "Authentication error occurred."

<details><summary>minion output/log</summary>

```
2023-06-01 12:54:41,749 [salt._logging.impl:1069][ERROR   ][3386] An un-handled exception was caught by Salt's global exception handler:
AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    sys.exit(salt_call())
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/scripts.py", line 444, in salt_call
    client.run()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cli/call.py", line 50, in run
    caller.run()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cli/caller.py", line 95, in run
    ret = self.call()
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/cli/caller.py", line 202, in call
    ret["return"] = self.minion.executors[fname](
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/executors/direct_call.py", line 10, in execute
    return func(*args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/saltutil.py", line 1680, in cmd
    fcn_ret = _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/saltutil.py", line 1632, in _exec
    for ret_comp in _cmd(**cmd_kwargs):
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/client/__init__.py", line 895, in cmd_iter
    pub_data = self.run_job(
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/client/__init__.py", line 387, in run_job
    pub_data = self.pub(
  File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/client/__init__.py", line 1933, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
```
</details>

<details><summary>master log</summary>

```
2023-06-01 12:54:41,745 [salt.auth        :343 ][WARNING ][2729] Master runs as 'salt', but user in payload is 'root'
2023-06-01 12:54:41,746 [salt.auth        :348 ][WARNING ][2729] Authentication failure of type "user" occurred.
2023-06-01 12:54:41,746 [salt.master      :2221][WARNING ][2729] Authentication failure of type "user" occurred.
```
</details>

**Setup**
To eliminate other possible causes, tested with simple setup:
Ubuntu LXC container with packages for salt-master & salt-minion and unchanged default config.

- [x] VM (Virtualbox, KVM, etc. please specify)
- [x] container (LXC)
- [x] onedir packaging

**Steps to Reproduce the behavior**
- ensure that "user: salt" is set in master config
- run: salt-call saltutil.cmd '*' fun='test.ping'

**Expected behavior**
In above example, test.ping should be run on given targets. As it does if master is run as root.

**Versions Report**
<details><summary>salt --versions-report</summary>

```yaml
Salt Version:
          Salt: 3006.1

Python Version:
        Python: 3.10.11 (main, May  5 2023, 02:31:54) [GCC 11.2.0]

Dependency Versions:
          cffi: 1.14.6
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.2
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.9.8
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 5.4.1
         PyZMQ: 23.2.0
        relenv: 0.12.3
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: ubuntu 20.04.6 focal
        locale: utf-8
       machine: x86_64
       release: 5.4.0-148-generic
        system: Linux
       version: Ubuntu 20.04.6 focal
```
</details>

**Comment by dmurphy18** [comment]
The problem occurs with a 3005.1 classic salt-master on RHEL 7 (running as default root) with a 3006.1 salt-minion on Ubuntu 20.04.  Noticed an oddity the command ```salt-call --local saltutil.cmd '*' fun='test.ping'``` should not reach out to the master, but the error code would suggest so.

**Comment by hurzhurz** [comment]
@dmurphy18 It is strange that you get an other error. I don't see 'salt.exceptions.SaltClientError'.
With 3006.1 and 3005.1 (onedir or classic), I always get 'salt.exceptions.AuthenticationError'.
And it is dependent on what user the master is running as.

<details><summary>minion output from 3005.1 classic</summary>

```
[ERROR   ] An un-handled exception was caught by Salt's global exception handler:
AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    load_entry_point('salt==3005.1', 'console_scripts', 'salt-call')()
  File "/usr/lib/python3/dist-packages/salt/scripts.py", line 441, in salt_call
    client.run()
  File "/usr/lib/python3/dist-packages/salt/cli/call.py", line 50, in run
    caller.run()
  File "/usr/lib/python3/dist-packages/salt/cli/caller.py", line 95, in run
    ret = self.call()
  File "/usr/lib/python3/dist-packages/salt/cli/caller.py", line 202, in call
    ret["return"] = self.minion.executors[fname](
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/executors/direct_call.py", line 10, in execute
    return func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/saltutil.py", line 1653, in cmd
    fcn_ret = _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/saltutil.py", line 1605, in _exec
    for ret_comp in _cmd(**cmd_kwargs):
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 897, in cmd_iter
    pub_data = self.run_job(
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 387, in run_job
    pub_data = self.pub(
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 1930, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    load_entry_point('salt==3005.1', 'console_scripts', 'salt-call')()
  File "/usr/lib/python3/dist-packages/salt/scripts.py", line 441, in salt_call
    client.run()
  File "/usr/lib/python3/dist-packages/salt/cli/call.py", line 50, in run
    caller.run()
  File "/usr/lib/python3/dist-packages/salt/cli/caller.py", line 95, in run
    ret = self.call()
  File "/usr/lib/python3/dist-packages/salt/cli/caller.py", line 202, in call
    ret["return"] = self.minion.executors[fname](
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/executors/direct_call.py", line 10, in execute
    return func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 149, in __call__
    return self.loader.run(run_func, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1228, in run
    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/loader/lazy.py", line 1243, in _run_as
    return _func_or_method(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/saltutil.py", line 1653, in cmd
    fcn_ret = _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/saltutil.py", line 1605, in _exec
    for ret_comp in _cmd(**cmd_kwargs):
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 897, in cmd_iter
    pub_data = self.run_job(
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 387, in run_job
    pub_data = self.pub(
  File "/usr/lib/python3/dist-packages/salt/client/__init__.py", line 1930, in pub
    raise AuthenticationError(err_msg)
salt.exceptions.AuthenticationError: Authentication error occurred.
```
</details>

Setting "ipc_mode: tcp" in /etc/salt/master and /etc/salt/minion doesn't make a difference, still salt.exceptions.AuthenticationError.

<details><summary>Here are the exact steps how I can reproduce this with a minimal setup in an lxc-container:</summary>

```
root@host:~# lxc-create -t download -n salttest -- --dist ubuntu --release focal --arch amd64
root@host:~# lxc-start -n salttest -d
root@host:~# lxc-attach -n salttest

root@salttest:~# echo 'deb [trusted=true arch=amd64] https://repo.saltproject.io/salt/py3/ubuntu/20.04/amd64/latest focal main' >> /etc/apt/sources.list
root@salttest:~# echo '127.0.1.1 salt' >> /etc/hosts
root@salttest:~# apt-get update && apt-get install salt-master salt-minion
root@salttest:~# systemctl start salt-master && systemctl start salt-minion
root@salttest:~# salt-key -A -y

# test with master running as user salt (3006 default)
root@salttest:~# salt-call test.ping                             # -> works
root@salttest:~# salt-call saltutil.cmd '*' fun='test.ping'      # -> salt.exceptions.AuthenticationError: Authentication error occurred.

# test with user root
root@salttest:~# sed -i '/^user:/s/salt/root/' /etc/salt/master
root@salttest:~# systemctl restart salt-master
root@salttest:~# salt-call test.ping                             # -> works
root@salttest:~# salt-call saltutil.cmd '*' fun='test.ping'      # -> works
```
</details>

Well, and while the general description of saltutil in its documentation doesn't suggest that it can be used for such things, saltutil.cmd looks to be something special in case it is used via an master-minion: "Assuming this minion is a master, execute a salt command"
So I consider it as a cleaner solution to achieve something like "salt xyz cmd.run 'salt * test.ping'".

And while it is easier to demonstrate the issue, I don't think it makes sense to use saltutil.cmd on the CLI. Also not with test.ping.
I actually use it in Jinja to gather data from multiple minions (with e.g. grains.get or file.read).

**Comment by hurzhurz** [comment]
Just did another test:
If I set up a second lxc-container with just a minion, connected to the master in the first lxc-container, and run the saltutil.cmd from there, I get the same error as you:
"salt.exceptions.SaltClientError: The salt master could not be contacted. Is master running?"

I think this is to be expected and using saltutil.cmd just makes sense on a minion that runs on the same host as the master.
So I think this is not related to my issue.

**Comment by dmurphy18** [comment]
@hurzhurz Actually it is related to the command that you are using, the ipc_mode is not 'tcp' and the master published ipc is not available and hence the error.  There is no authentication checking failing here, it is a pure check on the transport that you are using.  The default transport for a salt-minion utilizes zeromq, to utilize pure tcp as a transport method, you would need to change configuration settings, for example from /etc/salt/minion
```
418 # Windows platforms lack posix IPC and must rely on slower TCP based inter-
419 # process communications.  ipc_mode is set to 'tcp' on such systems.
420 #ipc_mode: ipc
421 
422 # Overwrite the default tcp ports used by the minion when ipc_mode is set to 'tcp'
423 #tcp_pub_port: 4510
424 #tcp_pull_port: 4511
```

Lastly, as stated in the documentation, ```The Saltutil module is used to manage the state of the salt minion itself. It is used to manage minion modules as well as automate updates to the salt minion.```

What you are doing, trying to ping other salt-minion's from a salt-minion using the saltutil command appears to be beyond the scope of what saltutil was intended for.

Your original contention was that this was an authentication error due to the addition of user, group salt in 3006.x, and as can be seen this is not the case, especially with 3005.1 where both master and minion are root.

Please consider closing this issue, since the code is working as designed, unless there is some other point you wish to raise.

**Comment by hurzhurz** [comment]
@dmurphy18  Well, to be honest, I don't know how the command exactly works internally and in which way the ipc_mode plays a role here.
I use the command on the master-minion where both, the master and the minion process, run next to each other on the same host. This might be somewhat special.

Actually I pretty much just want to do what the documentation says about [saltutil.cmd](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.saltutil.html#salt.modules.saltutil.cmd):
`Assuming this minion is a master, execute a salt command`
The limitation that it has to be a master, makes sense to me. And it actually looks like the minion needs to read the master config file for saltutil.cmd: https://github.com/saltstack/salt/blob/master/salt/modules/saltutil.py#L1684

As an example, I use it in a state that creates a file /etc/ssh/ssh_known_hosts on the master with the public ssh host keys of all minions.
For this file, I have a created a Jinja template where I gather the keys via something like this:
`{% set result = salt['saltutil.cmd'](tgt='*',fun='cmd.run',arg=['cat /etc/ssh/ssh_host_*_key.pub']) %}`
When the state is apply on the master, this <ins>works perfectly fine, if the salt-master process is running as user root</ins>.
But if it is running as user salt, what is now default, it doesn't: "Authentication error occurred" / "Master runs as 'salt', but user in payload is 'root'"

If saltutil.cmd has a different purpose... any idea what it is?

**Comment by dmurphy18** [comment]
@hurzhurz Running the master and minion process on the same machine is special, and allows the minion access to the master's publisher ipc.  The code was shown to you why ipc needs to be tcp, that is a configuration setting that was explained, don't know how I could make that clearer to you that it will not work using zeromq as the transport rather than pure tcp.

As stated, the saltutil module is designed for adjusting the minion, not to go pinging other minion's from a minion, that is typically done via a salt-master, unless you have a rare masterless environment, which you have not indicated.

Basically, using saltutil.cmd in a manner which it is not intended to being used.

Nevertheless, can duplicate on Master with local Minion, and the authentication checking is occurring since on the minion side, I am running as user sudo_david and the sudo checks are occurring which is trigger the other authentication checks when that information is passed to the master, hence this may be a matter of setting sudo rights correctly.

**Comment by dmurphy18** [comment]
Problem is in these lines https://github.com/saltstack/salt/blob/05d3295eba2f0a209a51d6afa8b0a30b1eb298df/salt/auth/__init__.py#L334-L337
We now have multiple keys to content with user's root and salt, we are only checking for user root's key, and hence changing the configuration setting for user to root, we get the match of keys, since only the one user in the system to check against.

We need to cycle through and allow for all the users in the master that need to be considered.
Need to check through code to see if other assumptions of a single user in the salt-master to check against.
Also need to check against draft PR https://github.com/saltstack/salt/pull/64194 has addressed this issue.

**Comment by dmurphy18** [comment]
Closing this since associated PR https://github.com/saltstack/salt/pull/64490 merged

**Comment by hurzhurz** [comment]
@dmurphy18 It looks like your change is not included in the 3006.2 release. Is this intended?
